core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistinctRowCount.java

```diff
@@ -227,6 +227,21 @@ public Double getDistinctRowCount(Values rel, RelMetadataQuery mq,
         return 1D;
       }
     }
+
+    // try to remove const columns from the group keys, as they do not
+    // affect the distinct row count
+    ImmutableBitSet nonConstCols = rel.getNonConstColumns(groupKey);
+    if (nonConstCols.cardinality() == 0) {
+      // all columns are constants, the distinct row count should be 1
+      return 1D;
+    }
+
+    if (nonConstCols.cardinality() < groupKey.cardinality()) {
+      // some const columns can be removed, call the method recursively
+      // with the trimmed columns
+      return getDistinctRowCount(rel, mq, nonConstCols, predicate);
+    }
```

Comment:
What would it return without your change?

Reply:
It would return the of distinct row count from the input, which is null for most cases.